frame/staking/src/voter_bags.rs

```diff
@@ -0,0 +1,897 @@
+	/// Get a bag by its upper vote weight.
+	pub fn get(bag_upper: VoteWeight) -> Option<Bag<T>> {
+		debug_assert!(
+			T::VoterBagThresholds::get().contains(&bag_upper) || bag_upper == VoteWeight::MAX,
+			"it is a logic error to attempt to get a bag which is not in the thresholds list"
+		);
+		crate::VoterBags::<T>::try_get(bag_upper).ok().map(|mut bag| {
+			bag.bag_upper = bag_upper;
+			bag
+		})
+	}
+
+	/// Get a bag by its upper vote weight or make it, appropriately initialized.
+	pub fn get_or_make(bag_upper: VoteWeight) -> Bag<T> {
+		Self::get(bag_upper).unwrap_or(Bag { bag_upper, ..Default::default() })
+	}
+
+	/// Put the bag back into storage.
+	pub fn put(self) {
+		crate::VoterBags::<T>::insert(self.bag_upper, self);
+	}
+
+	/// Get the head node in this bag.
+	pub fn head(&self) -> Option<Node<T>> {
+		self.head.as_ref().and_then(|id| Node::get(self.bag_upper, id))
+	}
+
+	/// Get the tail node in this bag.
+	pub fn tail(&self) -> Option<Node<T>> {
+		self.tail.as_ref().and_then(|id| Node::get(self.bag_upper, id))
+	}
+
+	/// Iterate over the nodes in this bag.
+	pub fn iter(&self) -> impl Iterator<Item = Node<T>> {
+		sp_std::iter::successors(self.head(), |prev| prev.next())
+	}
+
+	/// Insert a new voter into this bag.
+	///
+	/// This is private on purpose because it's naive: it doesn't check whether this is the
+	/// appropriate bag for this voter at all. Generally, use [`VoterList::insert`] instead.
+	///
+	/// Storage note: this modifies storage, but only for the nodes. You still need to call
+	/// `self.put()` after use.
+	fn insert(&mut self, voter: VoterOf<T>) {
+		self.insert_node(Node::<T> {
+			voter,
+			prev: None,
+			next: None,
+			bag_upper: self.bag_upper,
+		});
+	}
+
+	/// Insert a voter node into this bag.
+	///
+	/// This is private on purpose because it's naive; it doesn't check whether this is the
+	/// appropriate bag for this voter at all. Generally, use [`VoterList::insert`] instead.
+	///
+	/// Storage note: this modifies storage, but only for the node. You still need to call
+	/// `self.put()` after use.
+	fn insert_node(&mut self, mut node: Node<T>) {
+		let id = node.voter.id.clone();
+
+		node.prev = self.tail.clone();
+		node.next = None;
+		node.put();
+
+		// update the previous tail
+		if let Some(mut tail) = self.tail() {
+			tail.next = Some(id.clone());
+			tail.put();
+		}
+
+		// update the internal bag links
+		if self.head.is_none() {
+			self.head = Some(id.clone());
+		}
+		self.tail = Some(id.clone());
+
+		crate::VoterBagFor::<T>::insert(id, self.bag_upper);
+	}
+
+	/// Remove a voter node from this bag.
+	///
+	/// This is private on purpose because it doesn't check whether this bag contains the voter in
+	/// the first place. Generally, use [`VoterList::remove`] instead.
+	///
+	/// Storage note: this modifies storage, but only for adjacent nodes. You still need to call
+	/// `self.put()` and `node.put()` after use.
+	fn remove_node(&mut self, node: &Node<T>) {
+		node.excise();
```

Comment:
Don't we need to remove the node from storage? I don't see that being done anywhre explicitly. It gets removed from the list but afaict it will just stay in `VoterNodes`